include/llvm/Support/WinTypes.h

```diff
@@ -433,28 +433,69 @@ class CSimpleArray : public std::vector<T> {
 #define IsEqualIID(a, b) a==b
 #define IsEqualCLSID(a, b) !memcmp(&a, &b, sizeof(GUID))
 
-class CW2A {
-public:
-  CW2A(const wchar_t *wide, uint32_t utf = 0) {
-    (void)utf;
-    size_t len = wcslen(wide) + 1; // +1 for '\0'
-    m_psz = (char*) malloc(len * sizeof(char));
-    wcstombs(m_psz, wide, len);
-    m_psz[len-1] = '\0';
-  }
-  ~CW2A() {
-    free(m_psz);
+template< int t_nBufferLength = 128 >
+class CW2AEX
+{
+  public:
+  CW2AEX( LPCWSTR psz , UINT nCodePage = CP_UTF8)
+  {
+    if (nCodePage != CP_UTF8) {
+      // you have reached the limits of this half-assed implementation. So sorry.
+      m_psz = NULL;
+      return;
+    }
+
+    if (!psz) {
+      m_psz = NULL;
+      return;
+    }
+
+    int len = wcslen(psz)*4;
+    m_psz = new char[len];
+    std::wcstombs(m_psz, psz, len);
   }
-  const char* c_str() const { return m_psz; }
-  operator const char*() const { return m_psz; }
 
-public:
+  ~CW2AEX() { delete [] m_psz; }
+
+  operator LPSTR() const { return m_psz; }
+
   char *m_psz;
+  char m_szBuffer[t_nBufferLength]; // dunno what this is for
```

Comment:
Delete this line completely?

Reply:
The only reason for including that is it's part of the published interface of CW2AEX. Plus, if I remember my template edge cases correctly, compilers get cross if you don't use the template variable anywhere and if it's no longer a template, then it can't be used like CW2AEX<> as it is in various places.  I will certainly remove the comment and maybe explain this in a comment instead.

I will test the unused template variable though.